### DISCUSSION TITLE
rauc install for 110MB file takes 15 minutes

### DISCUSSION THREAD

**Discussion by formidable01** [rauc/rauc#1113] [Q&A]
Hi,

I have a bundle for a root filesystem that is about 100MB in size. When I install the new bundle, it takes about 15 minutes which seems awfully long to me.  I have two questions: 1) Why does it take so long? 2) How can I speed it up? I would note that when it completes, it all works just fine.  The issue is the execution time. Almost all of that 15 minutes is spent after the install process indicates that it is 90% complete

**Comment by a3f** [comment]
15 minutes is indeed awfully long. Can you provide some more info? 

- Are you using raw or verity bundles?
- What slot type do you have? (e.g. ext4, ubifs)?
- What image type is being flashed? (e.g. tar.gz, ext4)?
- What storage technology are you installing onto (eMMC? raw NAND?)
- What CPU does your system have?

With this information, it may be possible to ask more specific questions to see what you might need to configure differently.

**Reply by formidable01** [reply]
Thanks for the response.  Here is the additional info you requested.

1. I am using the default bundle which I believe is verity
2. ext4 slot
3. ext4 image
4. eMMC
5. I am using an Nvidia Orin system

Thanks again!

**Reply by jluebbe** [reply]
The default (if no format is specified in the bundle) is 'plain'. We'd suggest using the verity format instead.

**Comment by ejoerns** [comment]
> Almost all of that 15 minutes is spent after the install process indicates that it is 90% complete

This is just a result of bad progress value distribution which has been addressed in #808.

To get an idea of why it takes so long it would be interesting to compare the installation time with a simple copying of the image (e.g. via `dd`).

Do you install the bundle from network or from local storage? Do you use an 'adaptive update' method?

**Reply by formidable01** [reply]
I am installing the image from local store and I do not use adaptive update

**Reply by a3f** [reply]
- Can you compare against `dd` to the eMMC partition as @ejoerns suggested?
- What does `dmesg | grep 'device-mapper: verity:'` say?
- Do you have any `CRYPTO_SHA256_` enabled in Kconfig? e.g. `zcat /proc/config.gz | grep CRYPTO_SHA256`
- What is the output of `cat /sys/kernel/debug/mmc0/ios`? (Assuming `mmc0` is your eMMC)

**Reply by formidable01** [reply]
> * Can you compare against `dd` to the eMMC partition as @ejoerns suggested?

I think this identifies the issue.  Although the bundle compresses almost to nothing (I assume because the ext4 is mostly empty), the ext4 file itself is 30GB and although 15 minutes is a long time, it compares closely to the dd copy time. So the issue becomes one of how to manage the physical file size to reduce the copy time -or- to just live with the long copy time

> * What does `dmesg | grep 'device-mapper: verity:'` say?

`There is no such listing
`
> * Do you have any `CRYPTO_SHA256_` enabled in Kconfig? e.g. `zcat /proc/config.gz | grep CRYPTO_SHA256`

`Yes. These config items are enabled.`

> * What is the output of `cat /sys/kernel/debug/mmc0/ios`? (Assuming `mmc0` is your eMMC)

```
clock:          200000000 Hz
vdd:            21 (3.3 ~ 3.4 V)
bus mode:       2 (push-pull)
chip select:    0 (don't care)
power mode:     2 (on)
bus width:      3 (8 bits)
timing spec:    10 (mmc HS400 enhanced strobe)
signal voltage: 1 (1.80 V)
driver type:    0 (driver type B)
```

**Reply by ejoerns** [reply]
> I think this identifies the issue. Although the bundle compresses almost to nothing (I assume because the ext4 is mostly empty), the ext4 file itself is 30GB and although 15 minutes is a long time, it compares closely to the dd copy time. So the issue becomes one of how to manage the physical file size to reduce the copy time -or- to just live with the long copy time

I do not think that your ext4 has 30GiB of data. Thus the question is: Why is it that large? And the solution to your problem is: make it smaller.

How do you generate the ext4?

Note that RAUC can also automatically resize the written ext4 on the target if your really need that much space for additional things. But I'd assume that all the content written already comes with the bundle so that might not even be necessary.

**Reply by formidable01** [reply]
The ext4 file is built by yocto and perhaps sized to the eMMC (a guess).  If I take the yocto-generated ext4 file and use gzip on it to check how well it compresses, it gets down to around 140MB so I do in fact believe that the ext4 file is padded to fill out the target eMMC

**Reply by jluebbe** [reply]
I see two options:

- Use a tar archive instead of a ext4 image. This way, the filesystem is created by RAUC during installation just before the tar is extracted. With a tar, no empty space can exist.
- Reduce the size of the ext4 image generated by Yocto. Normally, Yocto tries to build an ext4 image that is just large enough for the installed packages. There are ways to customize this, of course: https://docs.yoctoproject.org/ref-manual/faq.html#how-do-i-create-images-with-more-free-space
  So you probably want to check where in your layers the default mechanism is overridden to generate such a large rootfs image.

**Reply by formidable01** [reply]
Thanks for the pointers. It is not clear to me how to get the process to use the tar archive instead of the ext4. Can you point me to the right place to implement that?

**Reply by jluebbe** [reply]
Hard to say without knowing how your image generation is configured. For your rootfs image, 
`IMAGE_FSTYPES` should contain tar, so you get a .tar in your deploy directory.

Then, in the bundle recipe, you can use `RAUC_IMAGE_FSTYPE` to select the tar instead of the raw ext4 image.

**Reply by formidable01** [reply]
I was reviewing the rauc code and as far as I can tell `RAUC_IMAGE_FSTYPE` can only be ext4 or raw . tar does not appear to be an accepted option. Of course, I could be wrong. I often am.

**Reply by ejoerns** [reply]
@formidable01 `RAUC_IMAGE_FSTYPE` is a bitbake variable. Where did you look it up in RAUC code?

**Reply by formidable01** [reply]
Yes, you are right. I was scouring the `meta-rauc/classes/bundle.bbclass` file and mistakenly referred to it as part of "rauc" instead of "bitbake". I remain confused about how to properly get the whole system to use tar.gz instead of ext4 to image the rootfs

**Reply by jluebbe** [reply]
This is more a general Yocto question, than RAUC. Take a look at:
https://docs.yoctoproject.org/bsp-guide/bsp.html?highlight=image_fstypes#bsp-machine-configuration-example

For standard images recipes such as `core-image-minimal`, that IMAGES_FSTYPES can cause a .tar to generated in your deploy dir. When that works, in a separate bundle recipe you use the bundle.bbclass option above to select which image needs to go into the bundle.

**Reply by formidable01** [reply]
Sorry, I appreciate your patience and perseverance with me. I can and do generate a tar.gz file from yocto but the problem is that `rauc install` doesn't appear to be able to install a tar.gz image. It can only install an ext4 image. If I execute `rauc install` with a bundle that has a tar,gz file in it - it doesn't work. What am I missing?

**Reply by jluebbe** [reply]
To install a tar.gz from a bundle to an ext4 slot, you need mkfs.ext4 on the target.

Do you get an error message (perhaps in the rauc.service journal output, if you use systemd)? See https://rauc.readthedocs.io/en/v1.6/using.html#debugging-rauc for some background.

Could you also post your `system.conf` and `rauc info <bundle>` somewhere? Perhaps as a https://gist.github.com/?

**Reply by formidable01** [reply]
Sorry for the extreme delay. Here are my files: https://gist.github.com/formidable01/a5558f7aa42cf14cd02439363f66aff5 I specify that they are ext4 because it is unclear to me that I am allowed to specify them as tar,gz  If I am then I suppose that solves my problem
